packages/@react-stately/color/src/useHexColorFieldState.ts

```diff
@@ -0,0 +1,133 @@
+import {useControlledState} from '@react-stately/utils';
+
+export interface HexColorFieldState extends NumberFieldState {
+  colorValue: Color,
+  setInputValue: (value: string) => void
+}
+
+export const defaultMinValue = '#000000';
+export const defaultMaxValue = '#FFFFFF';
+
+export function useHexColorFieldState(
+  props: HexColorFieldProps
+): HexColorFieldState {
+  const {
+    minValue = defaultMinValue,
+    maxValue = defaultMaxValue,
+    step = 1,
+    value,
+    defaultValue,
+    onChange,
+    validationState
+  } = props;
+
+  const minColor = Color.parse(minValue);
+  const maxColor = Color.parse(maxValue);
+  const minColorInt = minColor.toHexInt();
+  const maxColorInt = maxColor.toHexInt();
+
+  const clampColor = useCallback((value: ColorInput) => {
+    try {
+      const color = Color.parse(value);
+      const colorInt = color.toHexInt();
+      if (colorInt < minColorInt) { return minColor; }
+      if (colorInt > maxColorInt) { return maxColor; }
+      return color;
+    } catch (err) {
+      return undefined;
+    }
+  }, [minColorInt, maxColorInt]);
+
+  const initialValue = clampColor(value);
+  const initialDefaultValue = clampColor(defaultValue);
+  const [colorValue, setColorValue] = useControlledState<Color>(initialValue, initialDefaultValue, onChange);
+
+  const initialInputValue = (value || defaultValue) && colorValue ? colorValue.toString('hex') : '';
+  const [inputValue, setInputValue] = useState(initialInputValue);
+
+  const increment = () => {
+    setColorValue((previousValue: Color) => {
+      const colorInt = previousValue ? previousValue.toHexInt() : minColorInt;
+      const newValue = `#${Math.min(colorInt + step, maxColorInt).toString(16).padStart(6, '0')}`;
+      const newColor = new Color(newValue);
+      setInputValue(newColor.toString('hex'));
+      return newColor;
+    });
+  };
+
+  const incrementToMax = useCallback(() => {
+    if (maxValue) {
+      setColorValue(maxColor);
+      setInputValue(maxColor.toString('hex'));
+    }
+  }, [maxValue, maxColor, setColorValue, setInputValue]);
+
+  const decrement = () => {
+    setColorValue((previousValue: Color) => {
+      if (!previousValue) { return minColor; }
+      const colorInt = previousValue.toHexInt();
+      const newValue = `#${Math.max(colorInt - step, minColorInt).toString(16).padStart(6, '0')}`;
+      const newColor = new Color(newValue);
+      setInputValue(newColor.toString('hex'));
+      return newColor;
+    });
+  };
+
+  const decrementToMin = useCallback(() => {
+    if (minValue) {
+      setColorValue(minColor);
+      setInputValue(minColor.toString('hex'));
+    }
+  }, [minValue, minColor, setColorValue, setInputValue]);
+
+  const setFieldInputValue = (value: string) => {
+    setInputValue(value);
+    value = value.trim();
+    if (!value.length) { return; }
+    if (!value.startsWith('#')) {
+      value = `#${value}`;
+    }
+    try {
+      const color = clampColor(value);
```

Comment:
lets move the clamp into setColorValue, so above should be something along these lines

```
  const [colorValue, setColorValue] = useControlledState<Color>(initialValue, initialDefaultValue, onChange);
  let safelySetColorValue = (value: Color) => {
    let newColor = clamp(value, minColor, maxColor);
    setColorValue(newColor);
  }
```
and then you'd call safelySetColorValue here and in all the other places

Reply:
Hi @snowystinger, I'm not sure how we can use `safelySetColorValue` to replace all `setColorValue` calls in `useHexColorFieldState`. `setColorValue` allows either a `Color` object or a function, and we pass a function in most cases so we can check for color value equality before triggering `onChange` with a new object.

Reply:
Resolved. Irrelevant after specs changes removing props for custom min & max